executor/new_index_lookup_join.go

```diff
@@ -0,0 +1,534 @@
+}
+
+// NextChunk implements the Executor interface.
+func (e *NewIndexLookUpJoin) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	for {
+		err := e.prepareJoinResult(goCtx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if e.joinResult.NumRows() == 0 {
+			return nil
+		}
+		for e.joinResultCursor < e.joinResult.NumRows() {
+			if chk.NumRows() == e.maxChunkSize {
+				return nil
+			}
+			chk.AppendRow(0, e.joinResult.GetRow(e.joinResultCursor))
+			e.joinResultCursor++
+		}
+	}
+}
+
+// Next implements the Executor interface.
+// Even though we only support read children in chunk mode, but we are not sure if our parents
+// support chunk, so we have to implement Next.
+func (e *NewIndexLookUpJoin) Next(goCtx goctx.Context) (Row, error) {
+	for {
+		err := e.prepareJoinResult(goCtx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if e.joinResult.NumRows() == 0 {
+			return nil, nil
+		}
+		row := e.joinResult.GetRow(e.joinResultCursor)
+		datumRow := make(types.DatumRow, row.Len())
+		for i := 0; i < row.Len(); i++ {
+			datumRow[i] = row.GetDatum(i, e.schema.Columns[i].RetType)
+		}
+		e.joinResultCursor++
+		return datumRow, nil
+	}
+}
+
+func (e *NewIndexLookUpJoin) prepareJoinResult(goCtx goctx.Context) error {
+	if e.joinResultCursor < e.joinResult.NumRows() {
+		return nil
+	}
+	e.joinResult.Reset()
+	e.joinResultCursor = 0
+	for {
+		task, err := e.prepareTask(goCtx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if task == nil {
+			return nil
+		}
+		for task.cursor < task.outerResult.NumRows() {
+			e.lookUpMatchedInners(task, task.cursor)
+			outerRow := task.outerResult.GetRow(task.cursor)
+			task.cursor++
+			err = e.resultGenerator.emitToChunk(outerRow, task.matchedInners, e.joinResult)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if e.joinResult.NumRows() > 0 {
+				return nil
+			}
+		}
+	}
+}
+
+func (e *NewIndexLookUpJoin) prepareTask(goCtx goctx.Context) (*lookUpJoinTask, error) {
+	task := e.task
+	if task != nil && task.cursor < task.outerResult.NumRows() {
+		return task, nil
+	}
+	select {
+	case task = <-e.resultCh:
+	case <-goCtx.Done():
+		return nil, nil
+	}
+	if task == nil {
+		return nil, nil
+	}
+
+	select {
+	case err := <-task.doneCh:
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	case <-goCtx.Done():
+		return nil, nil
+	}
+	e.task = task
+	return task, nil
+}
+
+func (e *NewIndexLookUpJoin) lookUpMatchedInners(task *lookUpJoinTask, rowIdx int) {
+	outerKey := task.encodedLookUpKeys.GetRow(rowIdx).GetBytes(0)
+	innerPtrBytes := task.lookupMap.Get(outerKey)
+	task.matchedInners = task.matchedInners[:0]
+	for _, b := range innerPtrBytes {
+		ptr := *(*chunk.RowPtr)(unsafe.Pointer(&b[0]))
+		matchedInner := task.innerResult.GetRow(ptr)
+		task.matchedInners = append(task.matchedInners, matchedInner)
+	}
+}
+
+func (ow *outerWorker) run(goCtx goctx.Context, wg *sync.WaitGroup) {
+	defer func() {
+		terror.Log(ow.executor.Close())
```

Comment:
it's better to close `ow.executor` outside in `NewIndexLookUpJoin.Close()`

Reply:
The outer executor's ownership is transferred to outerWorker, it should be closed by outerWorker.

Reply:
I think It's better to be closed by the person who opened it. This makes our code easy to understand. By the way, if we close it in `NewIndexLookUpJoin .Close()` we can catch that error instead of log it.

Reply:
We can open the executor in `outerWorker`.
We always log the close error.